plugin/hooks.go

```diff
@@ -155,4 +156,10 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// NotificationWillBeSent is invoked before a notification is sent.
+	// To modify the notified user map, return the replacement map,
+	//
+	// Minimum server version: 5.20
+	NotificationWillBeSent(c *Context, post *model.Post, mentions *model.ExplicitMentions) *model.ExplicitMentions
```

Comment:
What happens if a pluin returns `nil`?

Reply:
This is handled in the `plugin/plugintest/hooks.go` file, [here](https://github.com/mattermost/mattermost-server/pull/13775/files#diff-eadcbf8ece04711fa69be6ded0a87732R163)  
If the call returns `nil`, the original `mentions` arguments is returned back.

Reply:
Are you sure? As far as I understand the code, this layer just make sure the results from the plugin as propagated to the `app` layer. There are other hooks that allow `nil` as a return value for there hooks.

Reply:
Well, this function `NotificationWillBeSent(c *Context, post *model.Post, mentions *model.ExplicitMentions) *model.ExplicitMentions` is to be implemented by the plugin.  
The implementation in `plugin/plugintest/hooks.go` is where it's actually called from the server and it does handle the case of a nil value iirc.

Reply:
I'm mostly concerned that this is an edge case that will cause problems with the code. Would you mind writing an integration test in https://github.com/mattermost/mattermost-server/tree/3e3fcdf84502ac0a4652e1d6e0459e67ed5ac188/app/plugin_api_tests to confirm the app runs fine if `nil` is returned? I suppose it has to be an `manual` one. See https://github.com/mattermost/mattermost-server/blob/3e3fcdf84502ac0a4652e1d6e0459e67ed5ac188/app/plugin_api_test.go#L827-L832 for more details.

Reply:
I'll first fix the `app/notification.go` issue and will look into this for another commit.

Reply:
Could you, just to double chat we catch this edge case, add a test to `app/plugin_hooks_test.go` where `NotificationWillBeSent` returns `nil`? Thanks